### ISSUE TITLE
delaunay_reduction enters endless loop with symprec=1e-6

### ISSUE THREAD

**Issue by aoterodelaroza** [atztogo/spglib#36]
Hi,

The following minimal python example:

cell = ([
(   51.6844120000000      ,    -0.669817000000000     ,    -1.16125000000000),
(   25.6065930000000     ,   7.75474900000000      ,  -0.900394000000000),
(   17.4497600000000     ,  -3.17743500000000     ,   9.50127300000000)],
[( 0.157242960000000      ,  0.489779070000000      ,  0.217297330000000     ),
(  0.842757010000000      ,  0.510220960000000      ,  0.782702660000000     ),
(  0.940642050000000      ,  0.786088570000000      ,  0.382604940000000     ),
(  5.935792000000000E-002 ,  0.213911460000000      ,  0.617395050000000     ),
(  0.836182950000000      ,  4.117084000000000E-002 ,  0.407814410000000     ),
(  0.163817020000000      ,  0.958829190000000      ,  0.592185580000000     ),
(  0.936557090000000      ,  1.535919000000000E-002 ,  0.277940800000000     ),
(  6.344287999999999E-002 ,  0.984640840000000      ,  0.722059190000000     ),
(  0.159434770000000      ,  0.701716630000000      ,  0.112066710000000     ),
(  0.840565200000000      ,  0.298283410000000      ,  0.887933280000000     ),
(  0.269116860000000      ,  0.441602290000000      ,  8.030411000000000E-002),
(  0.730883110000000      ,  0.558397740000000      ,  0.919695880000000     ),
(  0.662492490000000      ,  0.277869070000000      ,  0.542755420000000     ),
(  0.337507480000000      ,  0.722130960000000      ,  0.457244570000000     ),
(  0.846660530000000      ,  0.227220680000000      ,  0.305127260000000     ),
(  0.153339440000000      ,  0.772779350000000      ,  0.694872730000000     ),
(  0.287851950000000      ,  0.637887530000000      ,  0.962137870000000     ),
(  0.712148030000000      ,  0.362112500000000      ,  3.786212000000000E-002),
(  0.404809510000000      ,  0.423222380000000      ,  0.985479960000000     ),
(  0.595190470000000      ,  0.576777660000000      ,  1.452003000000000E-002),
(  0.377523010000000      ,  0.580599540000000      ,  0.754866700000000     ),
(  0.622476960000000      ,  0.419400490000000      ,  0.245133280000000     ),
(  0.187684980000000      ,  0.849938960000000      ,  2.253073000000000E-002),
(  0.812314990000000      ,  0.150061070000000      ,  0.977469260000000     ),
(  0.269956830000000      ,  0.224552470000000      ,  0.185390770000000     ),
(  0.730043140000000      ,  0.775447570000000      ,  0.814609210000000     ),
(  0.428983890000000      ,  0.996977650000000      ,  9.177415999999999E-002),
(  0.571016090000000      ,  3.022380000000000E-003 ,  0.908225830000000     ),
(  0.192383470000000      ,  0.235109790000000      ,  0.325125700000000     ),
(  0.807616500000000      ,  0.764890240000000      ,  0.674874290000000     ),
(  0.364671770000000      ,  0.995752520000000      ,  0.534136780000000     ),
(  0.635328210000000      ,  4.247510000000000E-003 ,  0.465863210000000     ),
(  0.359080190000000      ,  0.121890920000000      ,  0.500517080000000     ),
(  0.640919780000000      ,  0.878109110000000      ,  0.499482910000000     ),
(  0.167908900000000      ,  0.456216010000000      ,  0.613688080000000     ),
(  0.832091070000000      ,  0.543784030000000      ,  0.386311910000000     ),
(  0.977217410000000      ,  0.668080990000000      ,  0.761888480000000     ),
(  2.278256000000000E-002 ,  0.331919050000000      ,  0.238111510000000     ),
(  0.984390090000000      ,  0.537176160000000      ,  0.797492630000000     ),
(  1.560988000000000E-002 ,  0.462823870000000      ,  0.202507360000000     ),
(  0.175818390000000      ,  0.203798530000000      ,  0.686172830000000     ),
(  0.824181580000000      ,  0.796201510000000      ,  0.313827160000000     ),
(  0.516944350000000      ,  0.736159740000000      ,  0.437571590000000     ),
(  0.483055620000000      ,  0.263840290000000      ,  0.562428400000000     ),
(  0.507272290000000      ,  0.958949180000000      ,  0.379697050000000     ),
(  0.492727680000000      ,  4.105085000000000E-002 ,  0.620302940000000     ),
(  0.837032000000000      ,  0.702233620000000      ,  0.911647350000000     ),
(  0.162967980000000      ,  0.297766410000000      ,  8.835264000000000E-002),
(  0.177459220000000      ,  0.110823550000000      ,  0.710825860000000     ),
(  0.822540750000000      ,  0.889176490000000      ,  0.289174130000000     ),
(  0.768867770000000      ,  2.919315000000000E-002 ,  0.873636640000000     ),
(  0.231132210000000      ,  0.970806880000000      ,  0.126363350000000     ),
(  0.742938970000000      ,  0.170643170000000      ,  0.859415400000000     ),
(  0.257061000000000      ,  0.829356860000000      ,  0.140584590000000     ),
(  0.609403330000000      ,  0.195691500000000      ,  0.987174080000000     ),
(  0.390596640000000      ,  0.804308530000000      ,  1.282590000000000E-002),
(  0.473026490000000      ,  0.444890170000000      ,  4.795484000000000E-002),
(  0.526973480000000      ,  0.555109860000000      ,  0.952045150000000     ),
(  0.178753290000000      ,  0.566854810000000      ,  0.567912720000000     ),
(  0.821246680000000      ,  0.433145220000000      ,  0.432087270000000     ),
(  7.870151000000000E-002 ,  0.684254990000000      ,  0.751119340000000     ),
(  0.921298460000000      ,  0.315745040000000      ,  0.248880650000000     ),
(  0.343755080000000      ,  0.355559600000000      ,  0.450232070000000     ),
(  0.656244890000000      ,  0.644440430000000      ,  0.549767920000000     ),
(  0.120333580000000      ,  0.750636700000000      ,  0.468031870000000     ),
(  0.879666390000000      ,  0.249363330000000      ,  0.531968120000000     )],
[ 1 , 1 , 2 , 2 , 1 , 1 , 1 , 1 , 1 , 1 , 2 , 2 , 3 , 3 , 3 , 3 , 1 , 1 , 3 , 3 , 3 , 3 , 3 , 3 , 2 , 2 , 3 , 3 , 3 , 3 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 1 , 3 , 3 , 3 , 3 , 3 , 3 , 3 , 3 , 1 , 1 , 4 , 4 , 4 , 4 , 3 , 3 , 1 , 1 , 3 , 3 , 3 , 3 , 3 , 3 ])
dataset = get_symmetry(cell,symprec=1e-6)

enters an infinite loop. The reason can be traced down to the routine delaunay_reduce_basis in delaunay.c. The cutoff used for the scalar product in the Delaunay reduction algorithm is the square of 2.77e-7, which is dangerously close to the machine epsilon (~1e-16 for a double). The correct Delaunay-reduced lattice basis in this particular example has some scalar products that are zero. Roundoff errors cause that an extended basis with non-positive scalar products can never be found, hence the infinite loop. A possible solution would be to implement a maximum step condition in the loop (rather than a while(true)) that crashes with an error message about the unreasonable precision requirement. (However, note that symprec=1e-6 is not that much smaller than the default symprec=1e-5 used in the tests.)

**Comment by atztogo** [comment]
Thanks to your explaination, I didn't need to nail this problem down.
1. I replaced the while-loop by a for-loop with a maximum number (#define NUM_ATTEMPT 100).
2. symprec * symprec is replaced by symprec. For this delaunay reduction does not play a very important role in spglib, so I think it's OK. Avoiding rounding error is more important. 
3. I feel defining a tolerance for length, angle, and volume consistently is not easy for me. So at this moment, I gave up to send an error message about the unreasonable precision. But at least as I learned from you, I have to be more careful about rounding error.

I pushed these changes to the develop branch. Thanks @aoterodelaroza.
